### PR TITLE
WT-6448 RTS to read all the required pages as part of tree walk

### PR DESCRIPTION
**Pull request by kommiharibabu**
RTS used to read internal pages as part of tree walk and read
the leaf pages internally and evict them is causing the tree walk
to be restarted. Instead of inventing the tree walk restart logic
again in RTS, better to load all the RTS required pages as part
of tree walk.

### CODE REVIEW

**Comment by agorrod** [comment]
@keithbostic Could you take a look at this one today?

It pretty much follows one of your suggestions on the ticket. I see that there are still test failures, but it would be good if you can take a look and at least agree that the general solution makes sense to you.

File: src/txn/txn_rollback_to_stable.c [hunk 1]
```diff
@@ -975,15 +925,8 @@ __rollback_to_stable_btree(WT_SESSION_IMPL *session, wt_timestamp_t rollback_tim
     /* There is nothing to do on an empty tree. */
     if (btree->root.page == NULL)
         return (0);
-    /*
-     * Ensure the eviction server is out of the file - we don't want it messing with us. This step
-     * shouldn't be required, but it simplifies some of the reasoning about what state trees can be
-     * in.
-     */
-    WT_RET(__wt_evict_file_exclusive_on(session));
```
Comment (agorrod) [inline src/txn/txn_rollback_to_stable.c:983]: Why don't you do this step any more? Is it just an optimization? If so I'd prefer to do it later and keep this change conservative.

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:983]: No, it is not an optimization. With the change of reading all the required leaf pages into the cache to abort increases the cache usage and rollback to stable is not evicting any of the pages explicitly, so enabling eviction on the tree where the RTS is performing takes care of evicting the pages that are loaded by RTS operation.

or If we don't want the eviction to be enabled on the tree where rollback to stable is performing, we need to evict the leaf pages locally once the rollback to stable is finished.

File: src/txn/txn_rollback_to_stable.c [hunk 2]
```diff
@@ -804,32 +775,18 @@ __rollback_abort_newer_updates(
...
-        if (page == NULL) {
-            WT_RET(__wt_page_in(session, ref, 0));
-            local_read = true;
-        }
-        page = ref->page;
+    WT_ASSERT(session, ref->page != NULL);
+    page = ref->page;
+    if (!__wt_page_is_modified(page) &&
+      !__rollback_page_needs_abort(session, ref, rollback_timestamp)) {
+        __wt_verbose(session, WT_VERB_RTS, "%p: page skipped", (void *)ref);
```
Comment (bvpvamsikrishna) [inline src/txn/txn_rollback_to_stable.c:786]: Don't we need to increment the stat representing pages skipped ?

File: src/txn/txn_rollback_to_stable.c (hunk 1 above)
Comment (agorrod) [reply src/txn/txn_rollback_to_stable.c:983]: Ah, I see. That makes sense, it's a bit sad that we need to make a change this significant now.

File: src/txn/txn_rollback_to_stable.c (hunk 2 above)
Comment (keithbostic) [reply src/txn/txn_rollback_to_stable.c:786]: I think this is OK, I think the "skipped pages" stat is only intended for pages the tree-walk itself skips.

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -1259,7 +1203,8 @@ __rollback_to_stable_btree_apply(WT_SESSION_IMPL *session)
          * Truncate history store entries for the non-timestamped table.
          * Exceptions:
          * 1. Modified tree - Scenarios where the tree is never checkpointed lead to zero
-         * durable timestamp even they are timestamped tables. Until we have a special indication
+         * durable timestamp even they are timestamped tables. Until we have a special
+         * indication
```
Comment (keithbostic) [inline src/txn/txn_rollback_to_stable.c:1207]: Comment needs to be filled.

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Thanks, Fixed.

File: src/txn/txn_rollback_to_stable.c (hunk 2 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:786]: Yes, the current stat is to represent how many pages are skipped as part of tree walk, so no need to increment here.

**Pull request merged**